### PR TITLE
add bufr_query to GDASApp

### PR DESCRIPTION
**Pull request by CoryMartin-NOAA**
Add in the bufr_query repository to GDASApp to support direct ingest of BUFR files into JEDI-based applications.

### CODE REVIEW

**Comment by CoryMartin-NOAA** [comment]
Tested by running: `./bufr2netcdf.x /scratch1/NCEPDEV/global/glopara/dump/gdas.20240717/00/atmos/gdas.t00z.esmhs.tm00.bufr_d ../../sorc/bufr-query/test/testinput/bufrtest_mhs_mapping.yaml mhs.{splits/satId}.nc`
```
outputFile: mhs.{splits/satId}.nc
PRE-MAIN-INFO BufrParser: Parsing file /scratch1/NCEPDEV/global/glopara/dump/gdas.20240717/00/atmos/gdas.t00z.esmhs.tm00.bufr_d
PRE-MAIN-INFO Executing Queries
PRE-MAIN-WARNING Warning: Query String */CLAT did not apply to subset NC021036
PRE-MAIN-WARNING Warning: Query String */HOLS did not apply to subset NC021036
PRE-MAIN-WARNING Warning: Query String */CLON did not apply to subset NC021036
PRE-MAIN-WARNING Warning: Query String */BRITCSTC/TMBR did not apply to subset NC021036
PRE-MAIN-WARNING Warning: Query String */HMSL did not apply to subset NC021036
PRE-MAIN-WARNING Warning: Query String */LSQL did not apply to subset NC021036
PRE-MAIN-WARNING Warning: Query String */SIID did not apply to subset NC021036
PRE-MAIN-WARNING Warning: Query String */BRITCSTC/CHNM did not apply to subset NC021036
PRE-MAIN-INFO Building Bufr Data
PRE-MAIN-INFO Exporting Data
PRE-MAIN-INFO Finished [78.778s]
PRE-MAIN-WARNING Category (metop-a) was not found in file.
PRE-MAIN-WARNING Category (noaa-18) was not found in file.
```

Produced:
```
-rw-r--r-- 1 Cory.R.Martin da 8.8M Jul 17 20:23 mhs.noaa-19.nc
-rw-r--r-- 1 Cory.R.Martin da  17K Jul 17 20:23 mhs.noaa-18.nc
-rw-r--r-- 1 Cory.R.Martin da 6.0M Jul 17 20:23 mhs.metop-c.nc
-rw-r--r-- 1 Cory.R.Martin da 6.2M Jul 17 20:23 mhs.metop-b.nc
-rw-r--r-- 1 Cory.R.Martin da  17K Jul 17 20:23 mhs.metop-a.nc
```
```
ncdump -h mhs.metop-c.nc 
netcdf mhs.metop-c {
dimensions:
	Location = 820798 ;
variables:
	int Location(Location) ;
		Location:_FillValue = 2147483647 ;

// global attributes:
		:platformCommonName = "MHS" ;
		:platformLongDescription = "MTYP 021-027 PROCESSED MHS Tb (NOAA-18-19, METOP-1,2,3)" ;

group: MetaData {
  variables:
  	int64 dateTime(Location) ;
  		dateTime:_FillValue = 9223372036854775807LL ;
  		dateTime:long_name = "Datetime" ;
  		dateTime:units = "seconds since 1970-01-01T00:00:00Z" ;
  	int latitude(Location) ;
  		latitude:_FillValue = 2147483647 ;
  		latitude:long_name = "Latitude" ;
  		latitude:units = "degree_north" ;
  		latitude:valid_range = -90.f, 90.f ;
...
```

**Review by ADCollard — DISMISSED** [review]
Nice

**Review by rmclaren — APPROVED** [review]

File: .gitmodules [hunk 1]
```diff
@@ -81,3 +81,7 @@
   path = parm/jcb-algorithms
   url = https://github.com/noaa-emc/jcb-algorithms
   branch = develop
+[submodule "sorc/bufr-query"]
+  path = sorc/bufr-query
+  url = https://github.com/noaa-emc/bufr-query.git
+  branch = master
```
Comment (rmclaren) [inline .gitmodules]: `branch = develop`

**Review by rmclaren — CHANGES_REQUESTED** [review]

**Comment by rmclaren** [comment]
@CoryMartin-NOAA I'm working on rewriting the history of bufr-query so its in a funny state right now. The main branch should be the develop branch.

File: .gitmodules (hunk 1 above)
Comment (rmclaren) [reply .gitmodules]: @CoryMartin-NOAA I'm working on rewriting the history of bufr-query so its in a funny state right now. The main branch should be the develop branch.

**Comment by emilyhcliu** [comment]
@CoryMartin-NOAA FYI.  We are going to cut a tag from bufr-query for JCSDA to include bufr-query in the spack-stack-1.8.0 release which is targeted for end of Aug/early Sep.

Steve wants us to make sure with NCO that the bufr-query is acceptable in WCOSS-2.  The bufr-query depends on ecbuild, NetCDF, and BUFR libraries.  I do not see any reason that NCO would prohibit the installation of bufr-query on WCOSS-2.
Is there a procedure to follow for adding new software to the spack-stack for WCOSS-2?

**Review by emilyhcliu — DISMISSED** [review]

**Comment by CoryMartin-NOAA** [comment]
@emilyhcliu we don't have spack-stack on WCOSS2 (yet, but maybe eventually?). To add a new library, we will have to work with Jacob and Hang Lei, there is a process, which should be easier since this is an internal library.

File: .gitmodules (hunk 1 above)
Comment (CoryMartin-NOAA) [reply .gitmodules]: Ok thanks, just let me know when I can/should update the submodule hash and the .gitmodules file.

**Comment by emilyhcliu** [comment]
> @emilyhcliu we don't have spack-stack on WCOSS2 (yet, but maybe eventually?). To add a new library, we will have to work with Jacob and Hang Lei, there is a process, which should be easier since this is an internal library.

Do you think the following procedure looks right to you?
1.  Create a tag in bufr-query repository
2. JCSDA picks up the tag from GitHub, includes it in the spack-stack-1.8.0 and publishes it on JCSDA [spack-stack repository](https://github.com/JCSDA/spack-stack) 
3. NOAA picks up the spack-stack-1.8.0 and installs it on HPC machines.  

We need to work with Jacob and Hang Lei for item # 3 listed above, right?

**Comment by CoryMartin-NOAA** [comment]
For non-wcoss, I'm not even sure Jacob/Hang need to be involved. I think each HPC has someone who just installs the spack-stack on them, there is a team (Dom, Alex Richert, etc.). Jacob/Hang are only needed for WCOSS/NCO interaction

File: .gitmodules (hunk 1 above)
Comment (RussTreadon-NOAA) [reply .gitmodules]: @rmclaren , is it now safe to switch back to `develop`?

**Review by RussTreadon-NOAA — COMMENTED** [review]
Looks good to me.  Only one question about `.gitmodules`.

File: .gitmodules (hunk 1 above)
Comment (rmclaren) [reply .gitmodules]: Yes, you should be good to do that.

Comment (RussTreadon-NOAA) [reply .gitmodules]: Update to `develop` at  09a2fb9

**Comment by RussTreadon-NOAA** [comment]
09a2fb9 also updates the `jcb` hash to [`8269ce4`](https://github.com/NOAA-EMC/jcb/commit/8269ce474fe864fdd6dc51913ccce9daca630ac2).  PR #1230 and references therein explain why this is necessary.

**Comment by CoryMartin-NOAA** [comment]
I cannot approve since I was the author of the PR. @RussTreadon-NOAA also has contributed. So we probably need a 3rd party to approve the PR before this can be merged.

**Review by ADCollard — APPROVED** [review]

**Comment by CoryMartin-NOAA** [comment]
@rmclaren will need to approve since he previously requested changes

**Review by rmclaren — APPROVED** [review]

**Review by RussTreadon-NOAA — APPROVED** [review]
Approve

**Comment by RussTreadon-NOAA** [comment]
I don't think the Orion cron for orion-GW-RT is active any more.  PR #1226  has had an active [orion-GW-RT label](https://github.com/NOAA-EMC/GDASApp/pull/1226#event-13562642807) for 5 days.

**Comment by RussTreadon-NOAA** [comment]
The `test_gdasapp_soca_JGLOBAL_MARINE_BMAT` is due to the fact GDASApp is ahead of g-w.  `test/soca/gw/CMakeLists.txt` is configured to execute j-job `"JGLOBAL_MARINE_BMAT"`.   This j-job is not yet in g-w `develop`.  g-w PR [#2749](https://github.com/NOAA-EMC/global-workflow/pull/2749) adds `jobs/JGLOBAL_MARINE_BMAT`.   `test_gdasapp_soca_JGLOBAL_MARINE_BMAT` immediately fails upon start with the error
```
/var/spool/slurmd/job63799201/slurm_script: line 52: /scratch1/NCEPDEV/da/Russ.Treadon/git/global-workflow/bufr_query/so\
rc/gdas.cd/bundle/gdas/../../jobs/JGLOBAL_MARINE_BMAT: No such file or directory
+ slurm_script[1]: postamble slurm_script 1721734005 127
+ preamble.sh[70]: set +x
End slurm_script at 11:26:47 with error code 127 (time elapsed: 00:00:02)
```

The failures for the `JGDAS_GLOBAL_OCEAN` jobs may also be related to this disconnect.

What do you think @guillaumevernieres?

**Comment by guillaumevernieres** [comment]
> The `test_gdasapp_soca_JGLOBAL_MARINE_BMAT` is due to the fact GDASApp is ahead of g-w. `test/soca/gw/CMakeLists.txt` is configured to execute j-job `"JGLOBAL_MARINE_BMAT"`. This j-job is not yet in g-w `develop`. g-w PR [#2749](https://github.com/NOAA-EMC/global-workflow/pull/2749) adds `jobs/JGLOBAL_MARINE_BMAT`. `test_gdasapp_soca_JGLOBAL_MARINE_BMAT` immediately fails upon start with the error
> 
> ```
> /var/spool/slurmd/job63799201/slurm_script: line 52: /scratch1/NCEPDEV/da/Russ.Treadon/git/global-workflow/bufr_query/so\
> rc/gdas.cd/bundle/gdas/../../jobs/JGLOBAL_MARINE_BMAT: No such file or directory
> + slurm_script[1]: postamble slurm_script 1721734005 127
> + preamble.sh[70]: set +x
> End slurm_script at 11:26:47 with error code 127 (time elapsed: 00:00:02)
> ```
> 
> The failures for the `JGDAS_GLOBAL_OCEAN` jobs may also be related to this disconnect.
> 
> What do you think @guillaumevernieres?

@RussTreadon-NOAA , yes, ignore the marine tests for now.

**Comment by RussTreadon-NOAA** [comment]
Run g-w C96C48_ufs_hybatmDA CI on Hera using g-w `develop` at [`65a7ab7`](https://github.com/NOAA-EMC/global-workflow/commit/65a7ab75dc0e4baba06a02e11ed0455787056a68) and `feature/bufr-query` at 929e8e6.  All jobs successfully run to completion
```
Hera(hfe07):/scratch1/NCEPDEV/stmp2/Russ.Treadon/EXPDIR/prjedi_bq$ rocotostat -d prjedi_bq.db -w prjedi_bq.xml -c all -s
   CYCLE         STATE           ACTIVATED              DEACTIVATED
202402231800        Done    Jul 23 2024 13:27:32    Jul 23 2024 13:51:01
202402240000        Done    Jul 23 2024 13:27:32    Jul 23 2024 15:55:15
```

**Comment by RussTreadon-NOAA** [comment]
@CoryMartin-NOAA and @guillaumevernieres , this branch is ready for merger into GDASApp `develop`.  Is merging OK or will this PR mess up other GDASApp or g-w PRs?

**Comment by CoryMartin-NOAA** [comment]
@RussTreadon-NOAA as far as I know this shouldn't impact anything else, I'm good with merging if @guillaumevernieres is

**Comment by RussTreadon-NOAA** [comment]
Thank you, @guillaumevernieres .  That's a cool way to answer a question ... just go ahead and merge!

**Pull request merged**